clients/subscriber.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clients
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/conduitio-labs/conduit-connector-gcp-pubsub/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/gammazero/deque"
+	"github.com/google/uuid"
+)
+
+// Subscriber represents a struct with a GCP Pub/Sub client,
+// queues for messages, channels and a function to cancel the context.
+type Subscriber struct {
+	*pubSub
+
+	lock       sync.Mutex
+	msgDeque   *deque.Deque[*pubsub.Message]
+	ackDeque   *deque.Deque[*pubsub.Message]
+	errorCh    chan error
+	canceledCh chan struct{}
+	ctxCancel  context.CancelFunc
+}
+
+// NewSubscriber initializes a new subscriber client and starts receiving a messages to the queue.
+func NewSubscriber(ctx context.Context, cfg config.Source) (*Subscriber, error) {
+	ps, err := newClient(ctx, cfg.General)
+	if err != nil {
+		return nil, fmt.Errorf("new pubsub client: %w", err)
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+
+	sub := &Subscriber{
+		pubSub:     ps,
+		msgDeque:   deque.New[*pubsub.Message](),
+		ackDeque:   deque.New[*pubsub.Message](),
+		errorCh:    make(chan error),
+		canceledCh: make(chan struct{}),
+		ctxCancel:  cancel,
+	}
+
+	go func() {
+		s := sub.pubSub.client.Subscription(cfg.SubscriptionID)
+
+		// set MaxExtension less than 0, to not extend the ack deadline for each message
+		s.ReceiveSettings.MaxExtension = -1
+
+		if err = s.Receive(cctx,
+			func(_ context.Context, m *pubsub.Message) {
+				sub.lock.Lock()
+				defer sub.lock.Unlock()
+
+				sub.msgDeque.PushBack(m)
+			},
+		); err != nil {
+			close(sub.canceledCh)
+
+			sub.errorCh <- fmt.Errorf("subscription receive: %w", err)
+		}
+
+		sub.canceledCh <- struct{}{}
+	}()
+
+	return sub, nil
+}
+
+// Next returns the next record or an error.
+func (s *Subscriber) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case err := <-s.errorCh:
+		return sdk.Record{}, err
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	default:
+		s.lock.Lock()
+		defer s.lock.Unlock()
+
+		if s.msgDeque.Len() == 0 {
+			return sdk.Record{}, sdk.ErrBackoffRetry
+		}
+
+		msg := s.msgDeque.PopFront()
+		if msg == nil {
+			return sdk.Record{}, sdk.ErrBackoffRetry
+		}
+
+		s.ackDeque.PushBack(msg)
+
+		position, err := getBinaryUUID()
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("get position: %w", err)
+		}
+
+		return sdk.Record{
+			Position:  position,
+			Metadata:  msg.Attributes,
+			CreatedAt: msg.PublishTime,
+			Payload:   sdk.RawData(msg.Data),
+		}, nil
+	}
+}
+
+// Ack indicates successful processing of a Message passed.
+func (s *Subscriber) Ack(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		s.lock.Lock()
+		defer s.lock.Unlock()
+
+		if s.ackDeque.Len() == 0 {
+			return nil
+		}
```

Comment:
This indicates that a message was read but never added to the ackQueue, which looks to be a bug?

Reply:
Correct. Deleted it.